### PR TITLE
signing wrapper for cosign signer in pubtools-sign

### PR DESCRIPTION
**Pull request by midnightercz**
Commit includes implementation of signing wrapper for pubstools-sign-cosign to enable signing of containers with cosign. 
This commit also includes fix for build from scratch operation which was discovered during testing

### CODE REVIEW

File: pubtools/_quay/signer_wrapper.py [hunk 1]
```diff
@@ -358,6 +355,97 @@ def remove_signatures(
...
 
+class CosignSignerSettingsSchema(Schema):
+    """Validation schema for cosign signer settings."""
+
+    quay_namespace = fields.String(required=True)
+    quay_host = fields.String(required=True)
+    dest_quay_api_token = fields.String(required=True)
+
+
+class CosignSignerWrapper(SignerWrapper):
```
Comment (querti) [inline pubtools/_quay/signer_wrapper.py:366]: This wrapper doesn't create signatures, only removes them?

**Review by querti — CHANGES_REQUESTED** [review]

File: pubtools/_quay/iib_operations.py [hunk 2]
```diff
@@ -463,7 +463,11 @@ def task_iib_build_from_scratch(
         metadata={"tags": {target_settings["quay_operator_repository"]: [tag]}},
         repos={target_settings["quay_operator_repository"]: [tag]},
     )
-    outdated_manifests = item_processor.generate_existing_manifests_metadata(vitem)
+    existing_manifests = item_processor.generate_existing_manifests_metadata(vitem)
+    outdated_manifests = []
+    for ref, tag, man_arch_dig in existing_manifests:
```
[the remark this answers is not in the document]
Comment (midnightercz) [reply pubtools/_quay/iib_operations.py:468]: I discovered this problem during testing. I'll mention it in the commit message

File: pubtools/_quay/signer_wrapper.py [hunk 3]
```diff
@@ -358,6 +355,97 @@ def remove_signatures(
...
+        )
+        return [
+            (repository, e.split(":")[-1].replace(".sig", "").replace("-", ":"))
+            for e in existing_signatures[1]
+        ]
+
+    def _filter_to_remove(
+        self,
+        signatures: List[Tuple[str, str, str]],
+        _exclude: Optional[List[Tuple[str, str, str]]] = None,
```
[the remark this answers is not in the document]
Comment (midnightercz) [reply pubtools/_quay/signer_wrapper.py:390]: yeah, but it's not advised to have default lists as keywords variables

File: pubtools/_quay/signer_wrapper.py [hunk 4]
```diff
@@ -358,6 +355,97 @@ def remove_signatures(
...
+            signatures (List[Tuple[str, str, str]]): List of (digest, tag, repository)
+            tuples of signautres to remove.
+            _exclude (Optional[List[Tuple[str, str, str]]]): List of  (digest, tag, repository)
+            tuples of signautres to keep.
+        """
+        repos = list(set([x[2] for x in signatures]))
+        signatures_to_remove = [(x[2], x[0]) for x in signatures]
+        signatures_to_exclude = [(x[2], x[0]) for x in _exclude or []]
+        existing_signatures = sum(
+            run_in_parallel(self._list_signatures, [(repo,) for repo in repos]).values(), []
```
[the remark this answers is not in the document]
Comment (midnightercz) [reply pubtools/_quay/signer_wrapper.py:404]: it wraps the repo into tuple as function arguments are required to be iterable

File: pubtools/_quay/signer_wrapper.py (hunk 1 above)
Comment (midnightercz) [reply pubtools/_quay/signer_wrapper.py:366]: Signatures are created by sign_container method in the parent class

**Review by querti — APPROVED** [review]

**Pull request merged**